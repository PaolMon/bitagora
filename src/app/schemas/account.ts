export interface BitagoraAccountProps {
    infos: {
        myNeighbors: {
            address: Buffer,
            ipAddress: string,
            port: number
        }[],
        hostedFiles: {
            owner: Buffer,
            merkleRoot: Buffer
        }[],
    },
    digitalAsset: {
        pending: {
            fileName: string,
            merkleRoot: Buffer
        }[],
        allowed: {
            fileName: string,
            merkleRoot: Buffer,
            secret: string
        }[],
        myFiles: {
            fileName: string,
            merkleRoot: Buffer,
            secret:string
        }[],
    }
}

export const infosSchema = {
    type: 'object',
    properties: {
        myNeighbors: {
            fieldNumber: 1,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    address: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    ipAddress: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    port: {
                        dataType:'uint32',
                        fieldNumber: 3
                    }
                }
            }
        },
        hostedFiles: {
            fieldNumber: 2,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    owner: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 2
                    }
                }
            }
        }
    },
    default: {
        myNeighbors: [],
        hostedFiles: []
    }
}

export const digitalAssetAccountSchema = {
    type: 'object',
    properties: {
        myFiles: {
            fieldNumber: 1,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    secret: {
                        dataType:'string',
                        fieldNumber: 3
                    }
                }
            }
        },
        pending: {
            fieldNumber: 2,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    }
                }
            }
        },
        allowed: {
            fieldNumber: 3,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    secret: {
                        dataType:'string',
                        fieldNumber: 3
                    }
                }
            }
        }
    },
    default: {
        myFiles: [],
        pending: [],
        allowed: []
    }
}
