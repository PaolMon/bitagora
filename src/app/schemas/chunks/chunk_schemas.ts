export const registeredChunksSchema = {
	$id: 'lisk/chunks/registeredChunks',
	type: "object",
	required: ["chunks"],
	properties: {
        chunks: {
		type: "array",
		fieldNumber: 1,
		items: {
			type: 'object',
			required: ['owner', 'merkleRoot', 'hostedBy', 'allowedViewers'],
			properties: {
				owner: {
					dataType: 'bytes',
					fieldNumber: 1
				},
				merkleRoot: {
					dataType: 'bytes',
					fieldNumber: 2
				},
				hostedBy: {
					type: 'array',
					fieldNumber: 3,
					items: {
						dataType: 'bytes'
					}
				},
				requestedBy: {
					type: 'array',
					fieldNumber: 4,
					items: {
						type: 'object',
						properties: {
							address: {
								dataType: 'bytes',
								fieldNumber: 1
							},
							requestTransaction: {
								dataType: 'bytes',
								fieldNumber: 2
							},
							responseTransaction: {
								dataType: 'bytes',
								fieldNumber: 3
							},
							requestType: {
								dataType: 'string',
								fieldNumber: 4
							},
							status: {
								dataType: 'string',
								fieldNumber: 5
							}
						}
					}
				},
				allowedViewers: {
					type: 'array',
					fieldNumber: 5,
					items: {
						type: 'object',
						properties: {
							address: {
								dataType: 'bytes',
								fieldNumber: 1
							},
							secret: {
								dataType: 'string',
								fieldNumber: 2
							},
						}
					}
				}
			},
		}
	  }
	}
}