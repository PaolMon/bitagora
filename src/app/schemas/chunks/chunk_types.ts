export type registered_chunks = {
    chunks: chunk[]
}

export type chunk = {
    merkleRoot: Buffer,
    owner: Buffer,
    hostedBy: Buffer[],
	requestedBy: request_object[],
    allowedViewers: viewer[],
}

export type viewer = {
	address: Buffer,
	secret: string
}


export type request_object = {
	address: Buffer,
    requestTransaction: Buffer,
    responseTransaction: Buffer,
	requestType: string,
	status: string
}