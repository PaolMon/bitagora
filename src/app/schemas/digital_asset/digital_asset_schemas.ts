export const digitalAssetSchema = {
	$id: 'lisk/digital_asset/digitalAsset',
	type: 'object',
	required: ['owner', 'fileName', 'fileSize', 'fileHash', 'merkleRoot', 'merkleHeight', 'secret', 'transactionID', 'previousAssetReference'],
	properties: {
		owner: {
			dataType: 'bytes',
			fieldNumber: 1
		},
		fileName:{
			dataType: 'string',
			fieldNumber: 2
		},
		fileSize: {
			dataType: 'uint32',
			fieldNumber: 3
		},
		fileHash: {
			dataType: 'bytes',
			fieldNumber: 4
		},
		merkleRoot: {
			dataType: 'bytes',
			fieldNumber: 5
		},
		merkleHeight: {
			dataType: 'uint32',
			fieldNumber: 6
		},
		secret: {
			dataType: 'string',
			fieldNumber: 7
		},
		transactionID: {
			dataType: 'bytes',
			fieldNumber: 8
		},
		previousAssetReference: {
			dataType: 'bytes',
			fieldNumber: 9
		}
	},
};

export const registeredAssetsSchema = {
	$id: 'lisk/digital_asset/registeredAssets',
	type: "object",
	required: ["registeredAssets"],
	properties: {
	  registeredAssets: {
		type: "array",
		fieldNumber: 1,
		items: {
			type: 'object',
			required: ['owner', 'fileName', 'fileSize', 'fileHash', 'merkleRoot', 'merkleHeight', 'secret', 'transactionID', 'previousAssetReference'],
			properties: {
				owner: {
					dataType: 'bytes',
					fieldNumber: 1
				},
				fileName:{
					dataType: 'string',
					fieldNumber: 2
				},
				fileSize: {
					dataType: 'uint32',
					fieldNumber: 3
				},
				fileHash: {
					dataType: 'bytes',
					fieldNumber: 4
				},
				merkleRoot: {
					dataType: 'bytes',
					fieldNumber: 5
				},
				merkleHeight: {
					dataType: 'uint32',
					fieldNumber: 6
				},
				secret: {
					dataType: 'string',
					fieldNumber: 7
				},
				transactionID: {
					dataType: 'bytes',
					fieldNumber: 8
				},
				previousAssetReference: {
					dataType: 'bytes',
					fieldNumber: 9
				}
			},
		}
	  }
	}
	
};
