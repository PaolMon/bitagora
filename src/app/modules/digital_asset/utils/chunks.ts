import {  BaseModuleDataAccess, StateStore } from "lisk-sdk";
import { codec } from "lisk-sdk";
import { registeredChunksSchema } from "../../../schemas/chunks/chunk_schemas";
import { chunk, registered_chunks } from "../../../schemas/chunks/chunk_types";

export const CHAIN_STATE_CHUNKS = "digitalAsset:registeredChunks";

export const setNewChunk= async (stateStore: StateStore, asset: chunk) => {

    const chunks:chunk[] = await getAllChunks(stateStore);

    chunks.push(asset);

    const registeredChunks: registered_chunks = {chunks: chunks.sort((a, b) => a.merkleRoot.compare(b.merkleRoot))};

    await stateStore.chain.set(
        CHAIN_STATE_CHUNKS,
        codec.encode(registeredChunksSchema, registeredChunks)
    );
}

export const updateChunk =async (stateStore:StateStore, asset:chunk) => {
    const chunks:chunk[] = await getAllChunks(stateStore);

    const c_index: number = chunks.findIndex((t) => t.merkleRoot.equals(asset.merkleRoot));
    
    if (c_index < 0) {
        throw new Error("Asset not found");
    }

    chunks[c_index] = asset;

    const registeredChunks: registered_chunks = {chunks: chunks}

    await stateStore.chain.set(
        CHAIN_STATE_CHUNKS,
        codec.encode(registeredChunksSchema, registeredChunks)
    );
}

export const getChunkByMerkleRoot = async (stateStore: StateStore, merkleRoot: Buffer) => {
    const chunks:chunk[] = await getAllChunks(stateStore);

    const c_index: number = chunks.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (c_index < 0) {
        throw new Error("Asset not found");
    }

    return chunks[c_index];    
}

export const checkChunkExistenceByMerkleRoot = async (stateStore: StateStore, merkleRoot: Buffer): Promise<boolean> => {
    const chunks:chunk[] = await getAllChunks(stateStore);

    const c_index: number = chunks.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (c_index < 0) {
        return false;
    }

    return true;    
}

export const getAllChunks = async (stateStore: StateStore) => {
    const registeredChunksBuffer = await stateStore.chain.get(
        CHAIN_STATE_CHUNKS
    );

    if (!registeredChunksBuffer) {
        return [];
    }

    const registeredChunks = codec.decode<registered_chunks>(
        registeredChunksSchema,
        registeredChunksBuffer
    );

    return registeredChunks.chunks;
}

export const _getAllJSONChunks =async (dataAccess: BaseModuleDataAccess) => {
    const registeredChunksBuffer = await dataAccess.getChainState(
        CHAIN_STATE_CHUNKS
    );
    
    if (!registeredChunksBuffer) {
        return [];
    }
    
    const registeredChunks = codec.decode<registered_chunks>(
        registeredChunksSchema,
        registeredChunksBuffer
    );
    
      return codec.toJSON(registeredChunksSchema, registeredChunks);
}

