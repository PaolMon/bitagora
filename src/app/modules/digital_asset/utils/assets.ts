import { BaseModuleDataAccess, StateStore } from "lisk-sdk";
import { codec } from "lisk-sdk";
import { digitalAsset, registeredAssets } from "../../../schemas/digital_asset/digital_asset_types";
import { registeredAssetsSchema } from "../../../schemas/digital_asset/digital_asset_schemas";

export const CHAIN_STATE_DIGITAL_ASSETS = "digitalAsset:registeredAssets";


export const setNewAsset= async (stateStore: StateStore, asset: digitalAsset) => {

    const DAs:digitalAsset[] = await getAllAssets(stateStore);

    DAs.push(asset);

    const registeredAssets: registeredAssets = {registeredAssets: DAs.sort((a, b) => a.merkleRoot.compare(b.merkleRoot))};

    await stateStore.chain.set(
        CHAIN_STATE_DIGITAL_ASSETS,
        codec.encode(registeredAssetsSchema, registeredAssets)
    );
}

export const getAllAssets = async (stateStore: StateStore) => {
    const registeredAssetsBuffer = await stateStore.chain.get(
        CHAIN_STATE_DIGITAL_ASSETS
    );

    if (!registeredAssetsBuffer) {
        return [];
    }

    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );

    return registeredAssets.registeredAssets;
}

export const getAssetByMerkleRoot = async (stateStore: StateStore, merkleRoot: Buffer) => {
    const DAs:digitalAsset[] = await getAllAssets(stateStore);

    const DA_index: number = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (DA_index < 0) {
        throw new Error("Asset not found");
    }

    return DAs[DA_index];
}

/*
    THE FOLLOWING FUNCTIONS USE DATA ACCESS INSTEAD OF STATE STORE TO READ DATA
*/


export const _getAssetByMerkleRoot = async (dataAccess: BaseModuleDataAccess, merkleRoot: Buffer): Promise<digitalAsset> => {
    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        throw new Error('No Assets Found in stateStore');
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    const DAs = registeredAssets.registeredAssets;

    const DA_index: number = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (DA_index < 0) {
        throw new Error("Asset not found");
    }

    return DAs[DA_index];
}

export const _getAllJSONAssets =async (dataAccess: BaseModuleDataAccess) => {
    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        return [];
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    
      return codec.toJSON(registeredAssetsSchema, registeredAssets);
}

export const _getAssetHistoryByMerkleRoot = async (dataAccess: BaseModuleDataAccess, merkleRoot: Buffer) => {

    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        return [];
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    const DAs = registeredAssets.registeredAssets;

    let related_assets: digitalAsset[] = []
    let index = -1;
    let asset: digitalAsset;

    while(!merkleRoot.equals(Buffer.alloc(0))) {
        
        index = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

        if (index < 0) {
            throw new Error("Asset not found");
        }

        asset = DAs[index];
        related_assets.push(asset);

        merkleRoot = asset.previousAssetReference;
    }

    return codec.toJSON(registeredAssetsSchema, {registeredAssets: related_assets});
    //return related_assets;
}