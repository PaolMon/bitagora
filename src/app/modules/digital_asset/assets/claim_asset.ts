import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { digitalAsset } from '../../../schemas/digital_asset/digital_asset_types';
import { getAssetByMerkleRoot, setNewAsset } from '../utils/assets';
import { chunk } from '../../../schemas/chunks/chunk_types';
import { getChunkByMerkleRoot, setNewChunk } from '../utils/chunks';
import { BitagoraAccountProps } from '../../../schemas/account';


export type claim = {
	oldMerkleRoot: Buffer,
	newMerkleRoot: Buffer,
	newMerkleHeight: number,
	newHosts: Buffer[],
	newSecret: string
}

export class ClaimAsset extends BaseAsset {
	public name = 'claim';
  	public id = 3;

  	// Define schema for asset
	public schema = {
		$id: 'digitalAsset/claim-asset',
		title: 'ClaimAsset transaction asset for digitalAsset module',
		type: 'object',
		required: ['oldMerkleRoot','newMerkleRoot', 'newMerkleHeight', 'newHosts'], 
		properties: {
			oldMerkleRoot:{
				dataType: 'bytes',
				fieldNumber: 1
			},
			newMerkleRoot:{
				dataType: 'bytes',
				fieldNumber: 2
			},
			newMerkleHeight: {
				dataType: 'uint32',
				fieldNumber: 3
			},
			newHosts: {
				type: 'array',
				fieldNumber: 4,
				items: {
					dataType: 'bytes'
				}
			},
			newSecret: {
				dataType: 'string',
				fieldNumber: 5
			}
		},
	};

	public validate({ asset }: ValidateAssetContext<claim>): void {
		// Validate your asset
		if (asset.newMerkleHeight === 0 && asset.newHosts.length < 3) {
			throw new Error("If file has been splitted into chunks, it must be sent to at least 3 hosts");
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<claim>): Promise<void> {
		
		const senderAddress: Buffer = transaction.senderAddress;
		const old_chunk: chunk = await getChunkByMerkleRoot(stateStore, asset.oldMerkleRoot);

		if (old_chunk.owner.compare(senderAddress) != 0) {
			throw new Error("You are not allowed to claim this asset");
		}

		const digital_asset: digitalAsset = await getAssetByMerkleRoot(stateStore, asset.oldMerkleRoot);

		digital_asset.owner = senderAddress;
		digital_asset.merkleRoot = asset.newMerkleRoot;
		digital_asset.merkleHeight = asset.newMerkleHeight;
		digital_asset.previousAssetReference = asset.oldMerkleRoot;
		digital_asset.transactionID = transaction.id;
		digital_asset.secret = asset.newSecret;
	
		const chunk: chunk = {
			owner: senderAddress,
			merkleRoot: asset.newMerkleRoot,
			hostedBy: [],
			requestedBy: [],
			allowedViewers: []
		} 
		
		const senderAccount = await stateStore.account.get<BitagoraAccountProps>(senderAddress);
		
		senderAccount.digitalAsset.myFiles.push({fileName: digital_asset.fileName, merkleRoot: digital_asset.merkleRoot, secret: asset.newSecret})

		await setNewChunk(stateStore, chunk);

		await setNewAsset(stateStore, digital_asset);

		await stateStore.account.set(senderAccount.address, senderAccount).catch();

	}
}
