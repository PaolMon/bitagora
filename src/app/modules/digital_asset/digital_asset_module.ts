/*
 * LiskHQ/lisk-commander
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

/* eslint-disable class-methods-use-this */

import {
    AfterBlockApplyContext,


    AfterGenesisBlockApplyContext, BaseModule,


    BeforeBlockApplyContext, codec, TransactionApplyContext
} from 'lisk-sdk';
import { BitagoraAccountProps, digitalAssetAccountSchema } from '../../schemas/account';
import { digitalAssetSchema } from '../../schemas/digital_asset/digital_asset_schemas';
import { digitalAsset } from '../../schemas/digital_asset/digital_asset_types';
import { ClaimAsset } from "./assets/claim_asset";
import { CreateAsset } from "./assets/create_asset";
import { RequestAsset } from "./assets/request_asset";
import { ResponseAsset } from "./assets/response_asset";
import { _getAllJSONAssets, _getAssetByMerkleRoot, _getAssetHistoryByMerkleRoot } from './utils/assets';
import { _getAllJSONChunks } from './utils/chunks';

export class DigitalAssetModule extends BaseModule {
    public actions = {
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
        getAllAssets: async () => _getAllJSONAssets(this._dataAccess),
        getAllChunks: async () => _getAllJSONChunks(this._dataAccess),
        getAssetHistory:async (params: Record<string, unknown>) => {
            let { merkleRoot } = params;
            if (!Buffer.isBuffer(merkleRoot) && typeof merkleRoot === 'string') {
                merkleRoot = Buffer.from(merkleRoot, 'hex')
            } 
            return _getAssetHistoryByMerkleRoot(this._dataAccess, merkleRoot as Buffer);
        },

        getAsset:async (params:Record<string, unknown>): Promise<object> => {
            let { merkleRoot } = params;
            if (!Buffer.isBuffer(merkleRoot) && typeof merkleRoot === 'string') {
                merkleRoot = Buffer.from(merkleRoot, 'hex')
            } 
            const asset = await _getAssetByMerkleRoot(this._dataAccess, merkleRoot as Buffer);
            //return JSON.stringify(asset);
            return codec.toJSON(digitalAssetSchema, asset)
        },

        getAssetOwner: async(params: Record<string, unknown>): Promise<string> => {
            let { merkleRoot } = params;
            if (!Buffer.isBuffer(merkleRoot) && typeof merkleRoot === 'string') {
                merkleRoot = Buffer.from(merkleRoot, 'hex')
            } 
            const asset:digitalAsset = await _getAssetByMerkleRoot(this._dataAccess, merkleRoot as Buffer);
            return asset.owner.toString('hex');
        },

        getAccountAssets:async (params:Record<string, unknown>) => {
            let { address } = params;
            if (!Buffer.isBuffer(address) && typeof address === 'string') {
                address = Buffer.from(address, 'hex')
            } 
            const account = await this._dataAccess.getAccountByAddress<BitagoraAccountProps>(address as Buffer);
            return account.digitalAsset;
        },
    };
    public reducers = {
        // Example below
        // getBalance: async (
		// 	params: Record<string, unknown>,
		// 	stateStore: StateStore,
		// ): Promise<bigint> => {
		// 	const { address } = params;
		// 	if (!Buffer.isBuffer(address)) {
		// 		throw new Error('Address must be a buffer');
		// 	}
		// 	const account = await stateStore.account.getOrDefault<TokenAccount>(address);
		// 	return account.token.balance;
		// },
    };
    public name = 'digitalAsset';
    public accountSchema = digitalAssetAccountSchema;
    public transactionAssets = [new CreateAsset(), new RequestAsset(), new ResponseAsset(), new ClaimAsset()];
    public events = [
        // Example below
        // 'digitalAsset:newBlock',
    ];
    public id = 1001;

    // public constructor(genesisConfig: GenesisConfig) {
    //     super(genesisConfig);
    // }

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccoounts = genesisBlock.header.asset.accounts;
    }
}
