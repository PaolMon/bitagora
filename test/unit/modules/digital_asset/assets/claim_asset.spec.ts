import { claim, ClaimAsset } from '../../../../../src/app/modules/digital_asset/assets/claim_asset';
import { Account } from '@liskhq/lisk-chain';
import { ApplyAssetContext, codec, ReducerHandler, StateStore, testing } from 'lisk-sdk';
import { DigitalAssetModule } from '../../../../../src/app/modules/digital_asset/digital_asset_module';
import { BitagoraAccountProps } from '../../../../../src/app/schemas/account';
import { digitalAsset, registeredAssets, request_status, request_type } from '../../../../../src/app/schemas/digital_asset/digital_asset_types';
import { registeredAssetsSchema } from '../../../../../src/app/schemas/digital_asset/digital_asset_schemas';
import { chunk, registered_chunks } from '../../../../../src/app/schemas/chunks/chunk_types';
import { registeredChunksSchema } from '../../../../../src/app/schemas/chunks/chunk_schemas';
import { CHAIN_STATE_CHUNKS } from '../../../../../src/app/modules/digital_asset/utils/chunks';
import { CHAIN_STATE_DIGITAL_ASSETS } from '../../../../../src/app/modules/digital_asset/utils/assets';

describe('ClaimAsset', () => {
  let transactionAsset: ClaimAsset;

	beforeEach(() => {
		transactionAsset = new ClaimAsset();
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(3);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('claim');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		
		describe('schema validation', () => {
      		it.todo('should throw errors for invalid schema');
      		it.todo('should be ok for valid schema');
    	});
	});

	describe('apply', () => {
		let stateStore: StateStore;
		let reducerHandler: ReducerHandler;
		let old_ownerAccount: Account<BitagoraAccountProps>;
		let senderAccount: Account<BitagoraAccountProps>;
		let context: ApplyAssetContext<claim>;
		let test_registered_assets: registeredAssets;
		let test_registered_chunks: registered_chunks;

		beforeEach(() => {
			old_ownerAccount = testing.fixtures.createDefaultAccount<BitagoraAccountProps>([DigitalAssetModule]);
			old_ownerAccount.digitalAsset.myFiles = [
				{
					fileName: 'prova_4',
					merkleRoot: Buffer.alloc(7, 'prova_4'),
					secret: 'secret'
				}
			]

			
			senderAccount = testing.fixtures.createDefaultAccount<BitagoraAccountProps>([DigitalAssetModule]);
			senderAccount.digitalAsset.allowed = [
				{
					fileName: 'prova_1',
					merkleRoot: Buffer.alloc(7, 'prova_1'),
					secret: 'secret'
				}
			]
			
			test_registered_assets = 
			{
				registeredAssets: [
					{
						owner: old_ownerAccount.address,
						fileName: 'prova_1',
						fileSize: 1,
						fileHash: Buffer.alloc(7, 'prova_1'),
						merkleRoot: Buffer.alloc(7, 'prova_1'),
						merkleHeight: 0,
						secret: 'secret',
						transactionID: Buffer.alloc(3,'tx1'),
						previousAssetReference: Buffer.alloc(0)
					},
					{
						owner: Buffer.alloc(5,'jonny'),
						fileName: 'prova_2',
						fileSize: 1,
						fileHash: Buffer.alloc(7, 'prova_2'),
						merkleRoot: Buffer.alloc(7, 'prova_2'),
						merkleHeight: 0,
						secret: 'secret',
						transactionID: Buffer.alloc(3,'tx2'),
						previousAssetReference: Buffer.alloc(0)
					},
					{
						owner: Buffer.alloc(5,'jonny'),
						fileName: 'prova_3',
						fileSize: 1,
						fileHash: Buffer.alloc(7, 'prova_3'),
						merkleRoot: Buffer.alloc(7, 'prova_3'),
						merkleHeight: 0,
						secret: 'secret',
						transactionID: Buffer.alloc(3,'tx3'),
						previousAssetReference: Buffer.alloc(0)
					},
					{
						owner: old_ownerAccount.address,
						fileName: 'prova_4',
						fileSize: 1,
						fileHash: Buffer.alloc(7, 'prova_4'),
						merkleRoot: Buffer.alloc(7, 'prova_4'),
						merkleHeight: 0,
						secret: 'secret',
						transactionID: Buffer.alloc(3,'tx4'),
						previousAssetReference: Buffer.alloc(0)
					}
				]
			}

			test_registered_chunks = 
			{
				chunks: [
					{
						merkleRoot: Buffer.alloc(7, 'prova_1'),
						owner: old_ownerAccount.address,
						hostedBy: [],
						requestedBy: [{
							address: senderAccount.address,
							requestTransaction: Buffer.alloc(3, 'tx1'),
							responseTransaction: Buffer.alloc(3, 'tx0'),
							requestType: request_type.view_only,
							status: request_status.accepted
						}],
						allowedViewers: [{
							address:senderAccount.address,
							secret: 'secret_2'
						}],
					},
					{
						merkleRoot: Buffer.alloc(7, 'prova_2'),
						owner: senderAccount.address,
						hostedBy: [],
						requestedBy: [{
							address: senderAccount.address,
							requestTransaction: Buffer.alloc(3, 'tx2'),
							responseTransaction: Buffer.alloc(3, 'tx0'),
							requestType: request_type.ownership,
							status: request_status.accepted
						}],
						allowedViewers: [],
					},
					{
						merkleRoot: Buffer.alloc(7, 'prova_3'),
						owner: Buffer.alloc(5,'jonny'),
						hostedBy: [],
						requestedBy: [],
						allowedViewers: [],
					},
					{
						merkleRoot: Buffer.alloc(7, 'prova_4'),
						owner: old_ownerAccount.address,
						hostedBy: [],
						requestedBy: [],
						allowedViewers: [],
					}
				]
			}
		
			stateStore = new testing.mocks.StateStoreMock({
				accounts:  [old_ownerAccount, senderAccount],
				chain: {
					"digitalAsset:registeredAssets": codec.encode(registeredAssetsSchema, test_registered_assets),
					"digitalAsset:registeredChunks": codec.encode(registeredChunksSchema, test_registered_chunks)
				}
			});

			reducerHandler = testing.mocks.reducerHandlerMock;

			context = testing.createApplyAssetContext<claim>({
				stateStore,
				reducerHandler,
				asset: {
					oldMerkleRoot: Buffer.alloc(7, 'prova_2'),
					newMerkleRoot: Buffer.alloc(11, 'new_prova_2'),
					newMerkleHeight: 3,
					newHosts: [],
					newSecret: 'new_secret'
				},
				transaction: { senderAddress: senderAccount.address, nonce: BigInt(1) } as any,
			});
	
			jest.spyOn(stateStore.chain, 'get');
			jest.spyOn(stateStore.chain, 'set');
		});

    	describe('valid cases', () => {
      		it('should update the state store',async () => {
				  await transactionAsset.apply(context);
				  const new_digital_asset: digitalAsset = {
						owner: senderAccount.address,
						fileName: 'prova_2',
						fileSize: 1,
						fileHash: Buffer.alloc(7, 'prova_2'),
						merkleRoot: Buffer.alloc(11, 'new_prova_2'),
						merkleHeight: 3,
						secret: 'new_secret',
						transactionID: context.transaction.id,
						previousAssetReference: Buffer.alloc(7, 'prova_2')
				  };
				  const new_chunk: chunk = {
					merkleRoot: Buffer.alloc(11, 'new_prova_2'),
					owner: senderAccount.address,
					hostedBy: [],
					requestedBy: [],
					allowedViewers: [],
				  };
				  let _tra = test_registered_assets;
				  let _trc = test_registered_chunks;
				  _tra.registeredAssets.push(new_digital_asset);
				  _tra.registeredAssets.sort((a, b) => a.merkleRoot.compare(b.merkleRoot));
				  _trc.chunks.push(new_chunk);
				  _trc.chunks = _trc.chunks.sort((a, b) => a.merkleRoot.compare(b.merkleRoot));

				  expect(stateStore.chain.set).toHaveBeenCalledWith(
					  CHAIN_STATE_CHUNKS,
					  codec.encode(
						  	registeredChunksSchema,
							  _trc
						)
				  );

				  expect(stateStore.chain.set).toHaveBeenCalledWith(
					CHAIN_STATE_DIGITAL_ASSETS,
					codec.encode(
							registeredAssetsSchema,
							_tra
					  )
				)

			  });
    	});

    	describe('invalid cases', () => {
      		it.todo('should throw error');
    	});
	});
});
