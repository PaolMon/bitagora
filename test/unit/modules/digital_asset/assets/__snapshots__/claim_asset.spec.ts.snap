// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClaimAsset constructor should have valid schema 1`] = `
Object {
  "$id": "digitalAsset/claim-asset",
  "properties": Object {
    "newHosts": Object {
      "fieldNumber": 4,
      "items": Object {
        "dataType": "bytes",
      },
      "type": "array",
    },
    "newMerkleHeight": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "newMerkleRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 2,
    },
    "newSecret": Object {
      "dataType": "string",
      "fieldNumber": 5,
    },
    "oldMerkleRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "oldMerkleRoot",
    "newMerkleRoot",
    "newMerkleHeight",
    "newHosts",
  ],
  "title": "ClaimAsset transaction asset for digitalAsset module",
  "type": "object",
}
`;
